### ISSUE TITLE
False warning "value assigned to `a` is never read" when read

### ISSUE THREAD

**Issue by thiolliere** [rust-lang/rust#45963]
in this example:
```rust
fn foo() -> Result<(), ()> {
    Ok(())
}

fn main() {
    let mut a = None;
    loop {
        let b = foo();
        if b.is_ok() {
            a = Some(b);
            break;
        }
    }
    println!("{:?}", a.unwrap().unwrap());
}
```

output: (on stable rustc 1.21.0 and nightly rustc 1.23.0-nightly)
```
   Compiling ressai v0.1.0 (file:///home/thiolliere/developpement/ressai)
warning: value assigned to `a` is never read
 --> main.rs:6:9
  |
6 |     let mut a = None;
  |         ^^^^^
  |
  = note: #[warn(unused_assignments)] on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.60 secs
     Running `/home/thiolliere/developpement/ressai/target/debug/ressai`
()
```

**Comment by durka** [comment]
The warning is correct. Your loop doesn't exit until `foo` returns `Ok`, allowing `a` to be reassigned to `Some(b)`, and the original `None` is never read. You can replace `let mut a = None;` with simply `let a;`.

**Comment by thiolliere** [comment]
ah yes indeed, thanks. It was explicit but I though it means a in never read.
